src/Symfony/Component/Form/Extension/Validator/Constraints/FormValidator.php

```diff
@@ -65,6 +65,9 @@ public function validate($form, Constraint $formConstraint)
 
             if ($groups instanceof GroupSequence) {
                 $validator->atPath('data')->validate($form->getData(), $constraints, $groups);
+
+                $this->filterFormFieldsGroupSequenceViolations($groups);
```

Comment:
What about iterating the groups provided by the group sequence instead and check after each iteration whether or not the number of violations increased? We would then stop iterating once the number of violations changed.

With the current solution validators for constraints in "later" groups are still executed which they are not supposed to be.

Reply:
Yes, I thought about that.

Fields that are validated in `FromValidator` have a form field order that is not related to the group validation order. Thus, in general, at this stage, we know almost nothing about the need for validation of the field. We know nothing about any other form fields that have been or will be validated, as well as of their groups.

The second point is that if we try to filter constraints by group before validation, we must reproduce all the logic of the group sequence, which is incorporated in the validator itself and does not belong to `FromValidator` extension. But this is not an ultimate solution, because we do not know which field with which groups go further. And we must filter out violations after each verification anyway.

Only filtering violations is the lesser evil here.